worker-document/src/main/java/com/hpe/caf/worker/document/BulkDocumentMessageProcessor.java

```diff
@@ -97,17 +97,32 @@ public void processTasks() throws InterruptedException
             return;
         }
 
-        // Cycle around the tasks and set RESULT_SUCCESS responses on them
+        // Cycle around the tasks and set the responses on them
         for (final BulkDocumentTask bulkDocumentTask : bulkDocumentTasks) {
 
-            // Get the task object
+            // Get the task objects
             final AbstractTask documentWorkerTask = bulkDocumentTask.getDocumentWorkerTask();
+            final WorkerTask workerTask = bulkDocumentTask.getWorkerTask();
 
-            // Create the WorkerResponse object
-            final WorkerResponse workerResponse = documentWorkerTask.createWorkerResponse();
+            try {
+                // Raise the onAfterProcessDocument and onAfterProcessTask events
+                for (final Document document : bulkDocumentTask.getDocuments()) {
+                    documentWorkerTask.raiseAfterProcessDocumentEvent(document);
+                }
+
+                documentWorkerTask.raiseAfterProcessTaskEvent();
+
+                // Create the WorkerResponse object
+                final WorkerResponse workerResponse = documentWorkerTask.createWorkerResponse();
+
+                // Set the response on the WorkerTask object
+                workerTask.setResponse(workerResponse);
```

Comment:
Is setting the response part of processing the task? If so then the raiseAfterProcessTaskEvent should come after this I think.

Reply:
No it's not part of processing the task.  Setting the response is something done by the framework, whereas processing the task refers to processing by the worker code.  Besides if you raised the event after setting the response then any final manipulations made in the event handler would effectively be lost since the response would have already been set.